StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c

```diff
@@ -8,6 +8,326 @@
 
 #include <StandaloneMmIplPei.h>
 
+/**
+  Search all the available firmware volumes for MM Core driver.
+
+  @param  MmFvBase             Base address of FV which included MM Core driver.
+  @param  MmFvSize             Size of FV which included MM Core driver.
+  @param  MmCoreFileName       GUID of MM Core.
+  @param  MmCoreImageAddress   MM Core image address.
+
+  @retval EFI_SUCCESS          The specified FFS section was returned.
+  @retval EFI_NOT_FOUND        The specified FFS section could not be found.
+
+**/
+EFI_STATUS
+LocateMmFvForMmCore (
+  OUT EFI_PHYSICAL_ADDRESS  *MmFvBase,
+  OUT UINTN                 *MmFvSize,
+  OUT EFI_GUID              **MmCoreFileName,
+  OUT VOID                  **MmCoreImageAddress
+  )
+{
+  EFI_STATUS           Status;
+  UINTN                FvIndex;
+  EFI_PEI_FV_HANDLE    VolumeHandle;
+  EFI_PEI_FILE_HANDLE  FileHandle;
+  EFI_PE32_SECTION     *SectionData;
+  EFI_FV_INFO          VolumeInfo;
+
+  //
+  // Search all FV
+  //
+  VolumeHandle = NULL;
+  for (FvIndex = 0; ; FvIndex++) {
+    Status = PeiServicesFfsFindNextVolume (FvIndex, &VolumeHandle);
+    if (EFI_ERROR (Status)) {
+      break;
+    }
+
+    //
+    // Search PEIM FFS
+    //
+    FileHandle = NULL;
+    Status     = PeiServicesFfsFindNextFile (EFI_FV_FILETYPE_MM_CORE_STANDALONE, VolumeHandle, &FileHandle);
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    ASSERT (FileHandle != NULL);
+    if (FileHandle != NULL) {
+      *MmCoreFileName = &((EFI_FFS_FILE_HEADER *)FileHandle)->Name;
+    }
+
+    //
+    // Search Section
+    //
+    Status = PeiServicesFfsFindSectionData (EFI_SECTION_PE32, FileHandle, MmCoreImageAddress);
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    //
+    // Great!
+    //
+    SectionData = (EFI_PE32_SECTION *)((UINT8 *)*MmCoreImageAddress - sizeof (EFI_PE32_SECTION));
+    ASSERT (SectionData->Type == EFI_SECTION_PE32);
+
+    //
+    // This is SMM BFV
+    //
+    Status = PeiServicesFfsGetVolumeInfo (VolumeHandle, &VolumeInfo);
+    if (!EFI_ERROR (Status)) {
+      *MmFvBase = (EFI_PHYSICAL_ADDRESS)(UINTN)VolumeInfo.FvStart;
+      *MmFvSize = VolumeInfo.FvSize;
+      return EFI_SUCCESS;
+    } else {
+      return EFI_NOT_FOUND;
+    }
+  }
+
+  return EFI_NOT_FOUND;
+}
+
+/**
+  Find largest unallocated MMRAM in current MMRAM descriptor block
+
+  @param[in, out] LagestMmramRangeIndex  Lagest mmram range index.
+  @param[in]      CurrentBlock           Current MMRAM descriptor block.
+
+**/
+VOID
+FindLargestSmramRange (

```

Comment:
Better to have the name to be FindLargest**M**mramRange?

Reply:
Updated in new commit